File: Java/src/ContiguousSubArray.java
import java.io.*;
import java.util.*;
public class ContiguousSubArray {
    public static void main(String args[] ) throws Exception {

        Scanner reader = new Scanner(System.in);  // Reading from System.in
        System.out.println("Enter a number: ");
        int n = reader.nextInt();
        int arr[]=new int[n];
        for(int i=0; i<n; i++) {
            arr[i] = reader.nextInt();
        }

        System.out.println(MaxSubArraySum(arr));
    }

    static int MaxSubArraySum(int arr[]) {
        int size = arr.length;
        int max_sum = arr[0];
        int sum = arr[0];
        for(int i=1; i<size; i++) {
            sum += arr[i];
            if(sum < 0)
                sum = 0;
            if(max_sum < sum)
                max_sum = sum;
        }
        return max_sum;
    }
}

File: ruby/hacker_rank/contiguous_sub_array.rb
# arr = [-2, -3, 4, -1, -2, 1, 5, -3]
arr = [2,3,-5,2, -1]
len = arr.length
sum, max = arr[0], arr[0]
# for x in 0...len-3 do
#   sum = arr[x] + arr[x+1] + arr[x+2]
#   for y in (x+3)...len do
#     sum += arr[y]
#     max1 = sum if sum > max1
#   end
# end
# puts max1

for x in 1...len do
  sum += arr[x]
  sum = 0 if sum < 0
  max = sum if sum > max
end
puts max